drivers: sensor: paj7620: adding gesture sensor driver

paultimke opened the pull request:
Adding a driver and two sample applications (polling and trigger) for using the PAJ7620 gesture sensor.
The sensor has additional capabilities (besides gesture mode, it also has proximity and cursor modes), but since this is my first PR I didn't want it to get excessively large, so this only implements the basic gesture functionality.

I tested this using a nucleo f334r8 board.

paultimke committed:
drivers: sensor: paj7620: added driver

Added driver for the PAJ7620 gesture sensor. For now,
just added basic gesture mode, although sensor also
has other modes (proximity and cursor modes).

Signed-off-by: Paul Timke Contreras <ptimkec@live.com>

paultimke committed:
tests: drivers: build_all: sensors: added paj7620

Added paj7620 to the testcase so that it is in build.

Signed-off-by: Paul Timke Contreras <ptimkec@live.com>

paultimke committed:
samples: sensor: paj7620_gesture: added sample

Added a sample for using the PAJ7620 gesture sensor

Signed-off-by: Paul Timke Contreras <ptimkec@live.com>

paultimke commented on drivers/sensor/pixart/paj7620/Kconfig:
```diff
@@ -0,0 +1,61 @@
+if PAJ7620
+
+choice PAJ7620_TRIGGER_MODE
+	prompt "Trigger mode"
```
Thanks! I didn't know about that. I just addressed this and your other comments on my latest push.

fabiobaltieri commented on samples/sensor/gesture_polling/CMakeLists.txt:
```diff
@@ -0,0 +1,9 @@
+
+cmake_minimum_required(VERSION 3.20.0)
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
+project(gesture_polling)
```
the way the sample is named and placed in the tree makes it sound it's generic but it's really not, it should be named after the specific sensor it tests

paultimke commented:
Hello @fabiobaltieri, thanks for the review. All of your comments should be addressed by now. Please let me know if any more changes are needed :)

fabiobaltieri commented on drivers/sensor/pixart/paj7620/paj7620_trigger.c:
```diff
@@ -0,0 +1,132 @@
+	k_sem_init(&data->trig_sem, 0, K_SEM_MAX_LIMIT);
+	k_thread_create(&data->thread,
+			data->thread_stack,
+			CONFIG_PAJ7620_THREAD_STACK_SIZE,
```
use `K_KERNEL_STACK_SIZEOF(data->thread_stack)` instead of the option directly, vaguely recall that there can be some difference in some particular architecture/configuration, mainly it confuses coverity

paultimke replied:
@fabiobaltieri Sure, done. I also updated some paths in the README.rst I forgot to update in my latest push.

kartben commented:
Marking as tentative for 4.1 but will depend on @MaureenHelm 's availability to review

kartben commented on samples/sensor/paj7620_gesture_trig/prj.conf:
```diff
@@ -0,0 +1,6 @@
+CONFIG_STDOUT_CONSOLE=y
+CONFIG_I2C=y
+CONFIG_SENSOR=y
+CONFIG_PAJ7620=y
```
```suggestion
```

Already defaults to y when an okay paj7620 devicetree node is found

kartben commented on samples/sensor/paj7620_gesture_polling/prj.conf:
```diff
@@ -0,0 +1,4 @@
+CONFIG_STDOUT_CONSOLE=y
+CONFIG_I2C=y
+CONFIG_SENSOR=y
+CONFIG_PAJ7620=y
```
```suggestion
```

kartben commented on samples/sensor/paj7620_gesture_polling/sample.yaml:
```diff
@@ -0,0 +1,10 @@
+    platform_allow: nucleo_f334r8
+    depends_on:
+      - i2c
+      - paj7620
```
This tag doesn't exist. You should use a filter instead. Like this: https://github.com/zephyrproject-rtos/zephyr/blob/846b4687154af5f57ba5a32610ec868d5f4ae61d/samples/sensor/max6675/sample.yaml#L14

Same for the other sample

kartben commented:
@paultimke you now have a merge conflict, sorry!

paultimke commented:
Hello @kartben, just fixed the conflict :) Let me know if anything else is needed

yperess commented on drivers/sensor/pixart/paj7620/paj7620.c:
```diff
@@ -0,0 +1,412 @@
+		reg_addr = initial_register_array[i][0];
+		value = initial_register_array[i][1];
+
+		ret = paj7620_byte_write(dev, reg_addr, value);
```
This is a lot of writes byte by byte, does the chip not support a burst write?

paultimke replied:
From what I saw, the datasheet only listed single read, single write and burst read. However, I tried using a burst write and it seemed to work well, even on 400kbit/s I didn't see any issue. I'll be changing this to use a burst write instead if you agree.

yperess commented on drivers/sensor/pixart/paj7620/paj7620.c:
```diff
@@ -0,0 +1,412 @@
+
+	switch (gest_data_reg1) {
+	case PAJ7620_MASK_INT_FLAG_1_GES_RIGHT:
+		ret = paj7620_check_fwd_bkwd_gest(dev, PAJ7620_GES_RIGHT, result);
```
I'm a little confused. So to read a gesture we check the interrupt register first. Looking at the datasheet more than 1 big might be set, so a `switch` doesn't seem right here, I think this needs to be a sequence of `if`s:

```c
if (gest_data_reg1 & PAJ7620_MASK_INT_FLAG_1_GES_RIGHT != 0) {
    // Handle gesture right
}
if (gest data_reg1 & PAJ7620_MASK_INT_FLAG_1_GES_LEFT != 0) {
  // Handle gesture left
}
```

The second thing I seem to misunderstand is why are we sleeping so much in `paj7620_check_fwd_bkwd_gest`? If the interrupt bit is set, doesn't that mean that the gesture already happened? I don't think we need to wait for it to happen again.

paultimke replied:
Indeed, I guess I didn't see much of a problem with that at the start because of the way I was testing with my sample apps. On the polling sample, my wait time was 100ms, which is quite small to detect any other gesture in between reads, and each i2c read clears the register so I only got maximum one bit set each time. With the trigger sample, having more than 1 bit set is quite difficult because on each interrupt we read and hence clear the register.

I now see this is not the right approach. If we use larger delays in a polling application (especially if it's > 1 sec), then we may get those cases when we have more than 1 set bit, and comparing directly with the mask will lead to wrong results.

I agree with using the `if` statements and I think, instead of returning a single gesture to the application, like I was doing before, maybe I can just return the whole bitmask to the application and let it handle cases with multiple detected gestures as it wants. So, it could be something like this
```C
uint16_t result;
if ((gest_data_reg1 & PAJ7620_MASK_INT_FLAG_1_GES_RIGHT) != 0) {
    result |= PAJ7620_GES_RIGHT;
}
if ((gest_data_reg1 & PAJ7620_MASK_INT_FLAG_1_GES_LEFT) != 0) {
    result |= PAJ7620_GES_LEFT;
}
// etc
```
That would permeate towards the application level, and the API user would also have to AND the sensor_value with the mask.

Regarding your second point, I agree. I'll remove those delays. There are sometimes cases when after a FORWARD gesture, a DOWN or BACKWARD is falsely detected because of the way one moves the hand trying to exit the sensor's field of view, but I think this is also something the application level can handle if it needs/wants to and these delays don't even help much because the sensor keeps sensing anyways.

Please let me know if you agree with what I said about the first point. Thanks!

paultimke replied:
Although now that I think about it, if the users will still have to AND the sensor value, then I might as well just return the contents of the register with the gesture flags (maybe just clean the irrelevant flags from the 2nd interrupt register) instead of repeating the process with the 'if' statements and rebuilding the bitmask with the ORs. It's the same operation the application will have to do.

What do you think?

yperess replied:
I'm OK with you just returning the register value and documenting that the caller might get multiple gestures.

yperess commented on drivers/sensor/pixart/paj7620/paj7620.c:
```diff
@@ -0,0 +1,412 @@
+	}
+
+	/* Fetch gesture data */
+	ret = paj7620_read_gesture(dev, &detected_gesture);
```
I think this should just read the interrupt register to see what gestures happened in the past and return. It shouldn't try to read a new gesture exactly when fetching.

paultimke replied:
Sure this makes sense, if I understand correctly this means simply reading the register on `paj7620_sample_fetch` and leaving the handling of the flags (the `if`s checking against the gesture masks) to `paj7620_channel_get`?

yperess replied:
Yup

yperess commented on drivers/sensor/pixart/paj7620/paj7620.c:
```diff
@@ -0,0 +1,412 @@
+		ret = paj7620_set_sampling_rate(dev, val);
+		break;
+
+	case SENSOR_ATTR_PAJ7620_GESTURE_ENTRY_TIME:
+		data->gest_entry_time = val->val1;
+		break;
+
+	case SENSOR_ATTR_PAJ7620_GESTURE_EXIT_TIME:
+		data->gest_exit_time = val->val1;
+		break;
```
Shouldn't these modify the various GestureStartTh/GestureEndTh values?

paultimke replied:
From what I understand, the GestureStartTh and GestureEndTh registers from the datasheet are more about a threshold on the objet's size, and these attributes I had were to set the delay times I had above. However, since I'm removing those delays, then I guess it doesn't make much sense to have these attributes anymore. Please tell me if you agree.

yperess replied:
Makes sense, thanks for clarifying

yperess reviewed (changes requested):
I think I need some clarifications about what I'm understanding from the datasheet and what I'm seeing in the driver.

paultimke commented:
Hello @yperess, I tried to answer your questions above and I agree with your comments. I'll be pushing the changes I described soon. Please let me know if you have any further requests or questions :)

paultimke commented:
Hello @yperess  I just updated the PR with the changes we talked about. Mainly, now the register value is returned directly. This has been documented in the public sensor header for paj7620. The entry and exit delay times were also removed as discussed.

Regarding the burst write for the init settings, I did start seeing issues sometimes with this, and as the datasheet only explicitly says it supports single writes, then I believe it might be best to stick with that.

Also, I added a 200 microsecond delay after sending the first I2C write that wakes up the device. I noticed that, occasionally, the sensor didn’t respond when coming from a power cycle, but always seemed to work after resets or directly after flashing. Adding some time after waking the sensor up fixed the issue. I got this value experimentally: with 100us it still sometimes missed the next transactions, but 200us was pretty reliable.

yperess commented on drivers/sensor/pixart/paj7620/paj7620.h:
```diff
@@ -0,0 +1,73 @@
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/sensor.h>
+#include <zephyr/drivers/i2c.h>
+#if CONFIG_PAJ7620_TRIGGER
```
There's no need to add these ifs in the header, just include it.

yperess commented on drivers/sensor/pixart/paj7620/paj7620.h:
```diff
@@ -0,0 +1,73 @@
+#endif
+};
+
+#ifdef CONFIG_PAJ7620_TRIGGER
```
Same here, it's ok to define these and just not use them

yperess commented on drivers/sensor/pixart/paj7620/paj7620.c:
```diff
@@ -0,0 +1,295 @@
+	}
+
+#ifdef CONFIG_PAJ7620_TRIGGER
+	ret = paj7620_trigger_init(dev);
```
Please remove the conditional declaration in the header and convert this to

```c
if (IS_ENABLED(CONFIG_PAJ620_TRIGGER)) {
  ret = paj620_trigget_init(dev);
  if (ret < 0) {
    LOG_ERR("Failed to enable interrupts");
    return ret;
  }
}
```

yperess commented on drivers/sensor/pixart/paj7620/paj7620_trigger.c:
```diff
@@ -0,0 +1,132 @@
+	const struct paj7620_config *config = dev->config;
+	struct paj7620_data *data = dev->data;
+
+	data->dev = dev;
```
Check if `int_gpio` is not pointing to a null device. You might have the Kconfig enabled but not have the interrupt line.

paultimke replied:
Isn't the NULL check of the port of the `gpio_dt_spec` already done in the call to `gpio_is_ready_dt(&config->int_gpio)`?
Besides this I also added a NULL check on `int_gpio.port` on the trigger set function.

yperess commented on drivers/sensor/pixart/paj7620/paj7620_trigger.c:
```diff
@@ -0,0 +1,132 @@
+		return ret;
+	}
+
+	ret = gpio_pin_interrupt_configure_dt(&config->int_gpio, GPIO_INT_EDGE_FALLING);
```
Don't enable the interrupt here, do it in paj7620_trigger_set when you actually have an interrupt handler. Same when the handler passed is null (the user wants to stop listening to interrupts, disable the interrupt)

paultimke commented:
Hello @yperess, thank you for your review. I addressed your comments on my latest push, just had a question regarding the NULL check on the trigger_init function. I left that question directly on your comment.

paultimke commented:
Hello @kartben @yperess @MaureenHelm, is there anything else needed for this?

paultimke commented:
Hello, I had to rebase on latest main because of some issues with CI

faxe1008 commented on samples/sensor/paj7620_gesture_polling/sample.yaml:
```diff
@@ -0,0 +1,9 @@
+sample:
```
Make this one sample for the sensor with a Kconfig maybe and just add two tests?

faxe1008 commented:
Looks very good now, only thing is that imo you can condense the polling and trig sample into one.

ubieda commented on drivers/sensor/pixart/paj7620/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) Paul Timke <ptimkec@live.com>
```
Please add the year to the Copyright

ubieda commented on tests/drivers/build_all/sensor/i2c.dtsi:
```diff
@@ -1265,7 +1265,13 @@ test_i2c_bmm350: bmm350@ac {
 };
 
-test_i2c_ltr329: ltr329@aa {
+test_i2c_ltr329: ltr329@ad {
```
Thanks for fixing this, although it seems unrelated to the existing PR. Can you put it on a separate one-liner PR?

paultimke replied:
Sure thing

ubieda commented on drivers/sensor/pixart/paj7620/paj7620.c:
```diff
@@ -0,0 +1,295 @@
+	return 0;
+}
+
+static int paj7620_sample_fetch(const struct device *dev, enum sensor_channel chan)
+{
+	int ret;
+	struct paj7620_data *data = dev->data;
+	const struct paj7620_config *config = dev->config;
+	uint8_t gest_data[2];
+
+	if (chan != SENSOR_CHAN_ALL
+		&& ((enum sensor_channel_paj7620)chan != SENSOR_CHAN_PAJ7620_GESTURES)) {
+		return -ENOTSUP;
+	}
+
+	/* We read from REG_INT_FLAG_1 and REG_INT_FLAG_2 even on polling mode
+	 * (without using interrupts) because that's where the gesture
+	 * detection flags are set.
+	 * NOTE: A set bit means that the corresponding gesture has been detected
+	 */
+
+	ret = i2c_burst_read_dt(&config->i2c, PAJ7620_REG_INT_FLAG_1, gest_data, sizeof(gest_data));
+	if (ret < 0) {
+		LOG_ERR("Failed to read gesture data");
+		return ret;
+	}
+
+	data->gesture_flags = sys_get_le16(gest_data);
+
+	return 0;
+}
```
I'm concerned this attempts to force the fetch/get API to an event-driven pattern. Why not instead implement this as a streaming trigger?

ubieda reviewed (commented):
I'm thinking whether this sensor should rather be implemented using read/decode API to enable streaming mode.

paultimke commented:
> I'm thinking whether this sensor should rather be implemented using read/decode API to enable streaming mode.

Hello @ubieda, I agree it would be best to use the read/decode API with this sensor. However, should I undo all changes here using fetch/get and update to use read/decode? Otherwise if you see value in letting the fetch/get exist, maybe this PR could stay and the read/decode implementation can be handled in a separate PR. A Kconfig option could be used to select which interface to use for the driver.

ubieda commented:
Sounds good to me

kartben commented:
@paultimke this needs a rebase due to conflict in build_all's dtsi file. Thanks!

paultimke commented:
Hello @kartben, thanks! I just updated the PR to fix the conflict and add the 2025 year on the copyrights as per @ubieda's comment. The read/decode implemenation will remain pending for a later PR.

paultimke commented:
Hello @MaureenHelm, @kartben, is there something else I should fix?
